WorldCat Discovery Service.js

```diff
@@ -0,0 +1,291 @@
+	
+}
+/**
+ * RIS Scraper Function
+ *
+ */
+
+function scrape(risURL) {
+	
+	ZU.doGet(risURL, function (text) {
+		//Z.debug(text);
+		//conference proceedings exported as CONF, but fields match BOOK better
+		text = text.replace(/TY\s+-\s+CONF\s+[\s\S]+?\n\s*ER\s+-/g, function(m) {
+			return m.replace(/^TY\s+-\s+CONF\s*$/mg, 'TY  - BOOK')
+				//authors are actually editors
+				.replace(/^A1\s+-\s+/mg, 'A3  - ');
+		})
+		
+		Zotero.debug("Importing corrected RIS: \n" + text);
+		
+		var translator = Zotero.loadTranslator("import");
+		translator.setTranslator("32d59d2d-b65a-4da4-b0a3-bdd3cfb979e7");
+		translator.setString(text);
+		translator.setHandler("itemDone", function (obj, item) {
+			item.extra = undefined;
+			item.archive = undefined;
+			//clean up ISBNs
+			if (item.ISBN){
+				var ISBN = item.ISBN.split(/\s/);
+				var ISBNarray = [];
+				for (var i=0; i<ISBN.length; i++){
+					if (ZU.cleanISBN(ISBN[i])){
+						ISBNarray.push(ZU.cleanISBN(ISBN[i]));
+					} 
+				}
+				item.ISBN = ISBNarray.join(" ");
+			}
+			//remove space before colon
+			item.title = item.title.replace(/\s+:/, ":")
+			
+			//remove trailing colon after place
+			if (item.place){
+				item.place = item.place.replace(/:\s*$/, "");
+			}
+
+			//remove traling period after publication
+			
+			if (item.publicationTitle){
+				item.publicationTitle=item.publicationTitle.replace(/\.\s*$/, "");
+			}
+			//remove trailing commar after publisher
+			if (item.publisher){
+				item.publisher = item.publisher.replace(/,\s*$/, "");
+			}
+			//correct field mode for corporate authors
+			for (var i=0; i<item.creators.length; i++) {
+				if (!item.creators[i].firstName){
+					item.creators[i].fieldMode=1;
+				}
+			}
+			
+			//number of pages gets mapped to section???
+			if (item.section){
+				//extract possible roman numerals and number of pages without the p
+				var numPages = item.section.match(/(([lxiv]+,\s*)?\d+)\s*p/);
+				if (numPages) item.numPages = numPages[1];
+			}
+
+			item.complete();
+		});
+		translator.getTranslatorObject(function(trans) {
+			trans.options.defaultItemType = 'book'; //if not supplied, default to book
+			trans.options.typeMap = {'ELEC': 'book'}; //ebooks should be imported as books
+
+			trans.doImport();
+		});
+	});
+}
+
+function doWeb(doc, url) {
+	var results = getSearchResults(doc);
+	if(detectWeb(doc, url)=="multiple") {
+		var items = {};
+		for(var i=0, n=results.length; i<n; i++) {
+			var title = ZU.xpathText(results[i], './/div[@class="title"]');
+			//Z.debug(title)
+			if(!title) continue;
+			var oclcID= ZU.xpathText(results[i], './@data-oclcnum');
+			//Z.debug(oclcID)
+			var databaseID= ZU.xpathText(results[i], './@data-database-list');
+			//Z.debug(databaseID)
+			var risURL = composeURL(oclcID, databaseID);
+			Z.debug(risURL)
+			items[risURL] = title.trim();
+
+		}
+
+		Zotero.selectItems(items, function(items) {
+			if (!items) return true;
+
+			for (var i in items) {
+				scrape(i);
```

Comment:
I'd rather we collect these in an array and call `scrape` with that. The HTTP requests would then be sequential rather than concurrent, which is not very nice towards WorldCat.